rholang/src/main/scala/coop/rchain/rholang/interpreter/match.scala

```diff
@@ -332,4 +322,68 @@ object SpatialMatcher {
         } yield Unit
       case _ => StateT.liftF(None)
     }
+
+  def spatialMatch(target: Match, pattern: Match): OptionalFreeMap[Unit] =
+    for {
+      _ <- spatialMatch(target.target.get, pattern.target.get)
+      _ <- foldMatch(target.cases,
+                     pattern.cases,
+                     (t: MatchCase, p: MatchCase) => spatialMatch(t, p))
+    } yield Unit
+
+  /**
+    * Note that currently there should be no way to put a GPrivate in a pattern
+    * because patterns start with an empty environment.
+    * We're going to write the obvious definition anyway.
+    */
+  def spatialMatch(target: GPrivate, pattern: GPrivate): OptionalFreeMap[Unit] =
+    if (target == pattern)
+      StateT.pure(Unit)
+    else
+      StateT.liftF(None)
+
+  def spatialMatch(target: Channel, pattern: Channel): OptionalFreeMap[Unit] =
+    (target.channelInstance, pattern.channelInstance) match {
+      case (_, ChanVar(v)) if v.varInstance.isWildcard => StateT.pure(Unit)
+      case (Quote(p), ChanVar(v)) => {
+        v.varInstance match {
+          case FreeVar(level) => {
+            if (p.locallyFree.isEmpty)
+              StateT.modify(m => m + (level -> p))
+            else
+              StateT.liftF(None)
+          }
+          case _ => StateT.liftF(None)
+        }
+      }
+      case (ChanVar(tv), ChanVar(pv)) =>
+        (tv.varInstance, pv.varInstance) match {
+          case (BoundVar(tlevel), BoundVar(plevel)) => {
+            if (tlevel === plevel)
+              StateT.pure(Unit)
+            else
+              StateT.liftF(None)
+          }
+          case _ => StateT.liftF(None)
+        }
+      case (Quote(tproc), Quote(pproc)) => spatialMatch(tproc, pproc)
+      case _                            => StateT.liftF(None)
+    }
+
+  def spatialMatch(target: ReceiveBind, pattern: ReceiveBind): OptionalFreeMap[Unit] =
+    for {
+      _ <- spatialMatch(target.source.get, pattern.source.get)
+      _ <- {
+        if (target.patterns == pattern.patterns)
+          StateT.pure[Option, FreeMap, Unit](Unit)
+        else
+          StateT.liftF[Option, FreeMap, Unit](None)
+      }
+    } yield Unit
```

Comment:
Can we use the same programming pattern for MatchCase here?

```
  def spatialMatch(target: ReceiveBind, pattern: ReceiveBind): OptionalFreeMap[Unit] =
    if (target.patterns != pattern.patterns)
      StateT.liftF[Option, FreeMap, Unit](None)
    else
      spatialMatch(target.source.get, pattern.source.get)
```

Reply:
Or use the ReceiveBind programming pattern in the MatchCase if you prefer that.

Reply:
Yeah, we can use the early-out pattern here.